File: 67-add-binary/add-binary.cpp
class Solution {
public:
    string addBinary(string a, string b) {
        int carry = 0;
        string ans = "";
        int i = a.length()-1, j  = b.length()-1;
        while(i >= 0 && j >= 0){
            int cnt = 0;
            // string ai = (string)a[i];
            // string bj = b[j];
            ans = to_string(((int)a[i] - 48) ^ ((int)b[j] - 48) ^ carry) + ans;
            if(((int)a[i] - 48) == 1) cnt++;
            if(((int)b[j] - 48) == 1) cnt++;
            if(carry == 1) cnt++;
            if(cnt > 1) carry = 1;
            else carry = 0;
            i--;
            j--;
        }

        while(i >= 0){
            ans = to_string(((int)a[i] - 48) ^ carry) + ans;
            carry = ((int)a[i] - 48) & carry;
            i--;
        }

        while(j >= 0){
            ans = to_string(((int)b[j] - 48) ^ carry) + ans;
            carry = ((int)b[j] - 48) & carry;
            j--;
        }

        if(carry) ans = "1" + ans;

        return ans;
    }
};

File: 67-add-binary/add-binary.py
class Solution:
    def addBinary(self, a: str, b: str) -> str:

        i = len(a)-1
        j = len(b)-1

        res = ""
        carry = 0

        while(i >= 0 and j >= 0):
            total = int(a[i]) + int(b[j]) + carry
            if total <= 1:
                carry = 0
                res = str(total) + res
            elif total == 2:
                carry = 1
                res = "0" + res
            else:
                carry = 1
                res = "1" + res
            
            i -= 1
            j -= 1
        

        while(i >= 0):
            total = int(a[i]) + carry
            if total <= 1:
                carry = 0
                res = str(total) + res
            elif total == 2:
                carry = 1
                res = "0" + res
            i -= 1
        
        while(j >= 0):
            total = int(b[j]) + carry
            if total <= 1:
                carry = 0
                res = str(total) + res
            elif total == 2:
                carry = 1
                res = "0" + res
            j -= 1

        if carry == 1:
            res = str(carry) + res

        return res    
        